core/services/pipeline/runner.go

```diff
@@ -433,30 +444,74 @@ func (r *runner) ExecuteAndInsertFinishedRun(ctx context.Context, spec Spec, var
 		return 0, finalResult, nil
 	}
 
-	if runID, err = r.orm.InsertFinishedRun(r.orm.DB(), run, trrs, saveSuccessfulTaskRuns); err != nil {
+	if runID, err = r.orm.InsertFinishedRun(postgres.UnwrapGormDB(r.orm.DB()), run, saveSuccessfulTaskRuns); err != nil {
 		return runID, finalResult, errors.Wrapf(err, "error inserting finished results for spec ID %v", spec.ID)
 	}
 	return runID, finalResult, nil
 
 }
 
-func (r *runner) Run(ctx context.Context, run *Run, l logger.Logger, saveSuccessfulTaskRuns bool) (incomplete bool, err error) {
+func (r *runner) Run(ctx context.Context, run *Run, l logger.Logger, saveSuccessfulTaskRuns bool, fn func(tx *gorm.DB) error) (incomplete bool, err error) {
+	pipeline, err := r.initializePipeline(run)
+	if err != nil {
+		return false, err
+	}
+
+	async := pipeline.HasAsync()
+
+	err = postgres.GormTransactionWithDefaultContext(r.orm.DB(), func(tx *gorm.DB) error {
+		// avoid an extra db write if there is no async tasks present or if this is a resumed run
+		if async && run.ID == 0 {
+			now := time.Now()
+			// initialize certain task params
+			for _, task := range pipeline.Tasks {
+				switch task.Type() {
+				case TaskTypeETHTx:
+					run.PipelineTaskRuns = append(run.PipelineTaskRuns, TaskRun{
+						ID:            task.Base().uuid,
+						PipelineRunID: run.ID,
+						Type:          task.Type(),
+						Index:         task.OutputIndex(),
+						DotID:         task.DotID(),
+						CreatedAt:     now,
+					})
+				default:
+				}
+			}
+			if err = r.orm.CreateRun(postgres.UnwrapGorm(tx), run); err != nil {
+				return err
+			}
+
+		}
+
+		if fn != nil {
+			return fn(tx)
+		}
+		return nil
+	})
+	if err != nil {
+		return false, err
+	}
+
 	for {
-		trrs, err := r.run(ctx, run, NewVarsFrom(run.Inputs.Val.(map[string]interface{})), l)
+		_, err := r.run(ctx, pipeline, run, NewVarsFrom(run.Inputs.Val.(map[string]interface{})), l)
 		if err != nil {
 			return false, errors.Wrapf(err, "failed to run for spec ID %v", run.PipelineSpec.ID)
 		}
 
-		if run.Async {
-			var db *sql.DB
-			db, err = r.orm.DB().DB()
-			if err != nil {
-				return false, errors.Wrap(err, "unable to retrieve sql.DB")
-			}
-
+		if async {
 			var restart bool
 
-			restart, err = r.orm.StoreRun(db, run)
+			err = postgres.GormTransactionWithDefaultContext(r.orm.DB(), func(tx *gorm.DB) error {
+				restart, err = r.orm.StoreRun(postgres.UnwrapGorm(tx), run)
+				if err != nil {
+					return err
+				}
+				return nil
+			})
+
+			// TODO: how do we FailEarly with async since there's already state in the db? do we update the state or clear it out?
```

Comment:
What does FailEarly do?

Reply:
This was https://github.com/smartcontractkit/chainlink/pull/4715, it's intended for things like keeper checks, where it'll start the pipeline with a condition check that will keep failing when there's nothing interesting to do. In this case we want to fail immediately and terminate the run without bloating the db.

I confirmed with @spooktheducks today that we should go ahead and delete the rows when that happens.